sed/loader/sxp/loader.py

```diff
@@ -0,0 +1,891 @@
+# pylint: disable=duplicate-code
+"""
+This module implements the SXP data loader.
+This loader currently supports the SXP momentum microscope instrument.
+The raw hdf5 data is combined and saved into buffer files and loaded as a dask dataframe.
+The dataframe is a amalgamation of all h5 files for a combination of runs, where the NaNs are
+automatically forward filled across different files.
+This can then be saved as a parquet for out-of-sed processing and reread back to access other
+sed funtionality.
+Most of the structure is identical to the FLASH loader.
+"""
+import time
+from functools import reduce
+from pathlib import Path
+from typing import List
+from typing import Sequence
+from typing import Tuple
+from typing import Union
+
+import dask.dataframe as dd
+import h5py
+import numpy as np
+import pyarrow as pa
+from joblib import delayed
+from joblib import Parallel
+from natsort import natsorted
+from pandas import DataFrame
+from pandas import MultiIndex
+from pandas import Series
+
+from sed.core import dfops
+from sed.loader.base.loader import BaseLoader
+from sed.loader.sxp.metadata import MetadataRetriever
+from sed.loader.utils import parse_h5_keys
+from sed.loader.utils import split_dld_time_from_sector_id
+
+
+class SXPLoader(BaseLoader):
+    """
+    The class generates multiindexed multidimensional pandas dataframes from the new SXP
+    dataformat resolved by both macro and microbunches alongside electrons.
+    Only the read_dataframe (inherited and implemented) method is accessed by other modules.
+    """
+
+    __name__ = "sxp"
+
+    supported_file_types = ["h5"]
+
+    def __init__(self, config: dict) -> None:
+
+        super().__init__(config=config)
+        self.multi_index = ["trainId", "pulseId", "electronId"]
+        self.index_per_electron: MultiIndex = None
+        self.index_per_pulse: MultiIndex = None
+        self.failed_files_error: List[str] = []
+        self.nhits: np.ndarray = None
+
+    def initialize_paths(self) -> Tuple[List[Path], Path]:
+        """
+        Initializes the paths based on the configuration.
+
+        Returns:
+            Tuple[List[Path], Path]: A tuple containing a list of raw data directories
+            paths and the parquet data directory path.
+
+        Raises:
+            ValueError: If required values are missing from the configuration.
+            FileNotFoundError: If the raw data directories are not found.
+        """
+        # Parses to locate the raw beamtime directory from config file
+        if (
+            "paths" in self._config["core"]
+            and self._config["core"]["paths"].get("data_raw_dir", "")
```

Comment:
I guess this would also be useful for flashloader

Reply:
There are a few things that seem inconsistent to me still. E.g. I think it would be good if you could auto-find the beamtime-dirs, and still user-define a parquet-dir. But that's something for another PR, I think.

Reply:
I put this in #281